returnn/tf/layers/basic.py

```diff
@@ -1645,6 +1645,113 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
     return out
 
 
+class RandIntLayer(LayerBase):
+  """
+  Generates random numbers using ``tf.random.uniform``
+  """
+  layer_class = "rand_int"
+
+  # noinspection PyUnusedLocal
+  def __init__(self, shape, maxval, minval=0, dtype="int32", seed=None, **kwargs):
+    """
+    :param tuple[DimensionTag|int]|list[DimensionTag|int] shape: desired shape of output tensor
+    :param int maxval: upper bound (exclusive) on range of random values
+    :param int minval: lower bound (inclusive) on range of random values
+    :param str dtype: type of the output. For random ints, int32 and int64 make sense, but could also be floats
+    :param int|None seed: random seed
+    """
+    from returnn.tf.util.data import DimensionTag
+    super(RandIntLayer, self).__init__(**kwargs)
+
+    seed = seed if seed is not None else self.network.random.randint(2 ** 31)
+
+    shape_parsed, _ = self._parse_shape(shape)
+    shape_parsed = tuple(s if not isinstance(s, DimensionTag) else self.get_batch_dim() for s in shape_parsed)
+
+    self.output.placeholder = tf.random.uniform(shape_parsed, minval=minval, maxval=maxval, dtype=dtype, seed=seed)
+
+  @classmethod
+  def transform_config_dict(cls, d, network, get_layer):
+    """
+    :param dict[str] d:
+    :param returnn.tf.network.TFNetwork network:
+    :param (str)->LayerBase get_layer:
+    """
+    d.setdefault("from", [])
+    super(RandIntLayer, cls).transform_config_dict(d, network=network, get_layer=get_layer)
+
+  @classmethod
+  def _parse_shape(cls, shape):
+    """
+    Parse the desired shape for further usage in init and get_out_data_from_opts.
+    The returned shape includes batch dim as DimensionTag and dynamic axes as tensors.
+    Additionally, return dict of dynamic axes and their sizes.
+
+    :param tuple[DimensionTag|int]|list[DimensionTag|int] shape: desired shape of output tensor
+    :return: parsed shape including batch dim and sizes of the dynamic axes
+    :rtype: tuple[tuple|dict]
+    """
+    from returnn.tf.util.data import DimensionTag
+
+    shape_parsed = []
+    batch_dim_axis = None
+    dyn_axes_sizes = {}
+    for ax, s in enumerate(shape):
+      if isinstance(s, int):
+        shape_parsed.append(s)
+      elif isinstance(s, DimensionTag):
+        if s.kind == DimensionTag.Types.Batch:
+          assert batch_dim_axis is None
+          batch_dim_axis = ax
+          shape_parsed.append(s)
+        elif isinstance(s.dimension, int):
+          shape_parsed.append(s.dimension)
+        elif s.dimension is None:
+          assert s.dyn_size is not None
+          shape_parsed.append(tf.reduce_max(s.dyn_size))
```

Comment:
There should be no tensor operations in `get_out_data_from_opts`.
I'm not sure it is helpful to have this helper function for both `get_out_data_from_opts` and `__init__`. At least this makes it quite complicated.